extension/partiql-extension-ion/src/encode.rs

```diff
@@ -0,0 +1,417 @@
+use delegate::delegate;
+
+use ion_rs::{IonError, IonType, IonWriter};
+use ordered_float::OrderedFloat;
+use partiql_value::{Bag, DateTime, List, Tuple, Value};
+use rust_decimal::Decimal;
+use std::io::Write;
+
+use crate::common::{
+    BAG_ANNOT, DATE_ANNOT, MISSING_ANNOT, TIME_ANNOT, TIME_PART_HOUR_KEY, TIME_PART_MINUTE_KEY,
+    TIME_PART_SECOND_KEY, TIME_PART_TZ_HOUR_KEY, TIME_PART_TZ_MINUTE_KEY,
+};
+use crate::Encoding;
+use thiserror::Error;
+use time::{Date, Duration, Time, UtcOffset};
+
+/// Errors in ion encoding.
+///
+/// ### Notes
+/// This is marked `#[non_exhaustive]`, to reserve the right to add more variants in the future.
+#[derive(Error, Debug, Clone, PartialEq)]
+#[non_exhaustive]
+pub enum IonEncodeError {
+    /// Ion Writer error.
+    #[error("Ion write error: `{}`", .0)]
+    IonWriterError(IonError),
+
+    /// Unsupported type error.
+    #[error("Ion write error: unsupported value of type `{}`", .0)]
+    UnsupportedType(&'static str),
+}
+
+impl From<IonError> for IonEncodeError {
+    fn from(value: IonError) -> Self {
+        IonEncodeError::IonWriterError(value)
+    }
+}
+
+/// Result of attempts to encode a [`Value`] to Ion.
+pub type IonEncodeResult = Result<(), IonEncodeError>;
+
+/// Config for construction an Ion encoder.
+pub struct IonEncoderConfig {
+    mode: Encoding,
+}
+
+impl IonEncoderConfig {
+    /// Set the mode to `mode`
+    pub fn with_mode(mut self, mode: crate::Encoding) -> Self {
+        self.mode = mode;
+        self
+    }
+}
+
+impl Default for IonEncoderConfig {
+    fn default() -> Self {
+        IonEncoderConfig {
+            mode: crate::Encoding::Ion,
+        }
+    }
+}
+
+/// Builder for creating an encoder.
+pub struct IonEncoderBuilder {
+    config: IonEncoderConfig,
+}
+
+impl IonEncoderBuilder {
+    /// Create the builder from 'config'
+    pub fn new(config: IonEncoderConfig) -> Self {
+        Self { config }
+    }
+
+    /// Create a encoder given the previously specified config and an ion [`IonWriter`].
+    pub fn build<'a, W, I>(
+        self,
+        writer: &'a mut I,
+    ) -> Result<Box<dyn ValueEncoder<W, I> + 'a>, IonEncodeError>
```

Comment:
Similar to comment about the Ion decoder, is the result type necessary?

Reply:
You're right in that it is currently infallible. I modeled it with a result because I'm not certain that it will remain so if additional config options are added...